File: scripts/Binary Tree Inorder Traversal/Binary Tree Inorder Traversal.java
class Solution {
    public List<Integer> vals = new ArrayList<>();
    public List<Integer> inorderTraversal(TreeNode root) {
        if(root==null) return vals;
        inorderTraversal(root.left);
        vals.add(root.val);
        inorderTraversal(root.right);
        return vals;
    }
}

File: algorithms/Binary Tree Inorder Traversal/Binary Tree Inorder Traversal.py
from typing import List, Optional


class Solution:
	"""
	Time:   O(n)
	Memory: O(n)
	"""

	def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
		inorder = []
		stack = []

		while stack or root is not None:
			if root:
				stack.append(root)
				root = root.left
			else:
				node = stack.pop()
				inorder.append(node.val)
				root = node.right

		return inorder


class Solution:
	"""
	Time:   O(n)
	Memory: O(n)
	"""

	def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
		return list(self.inorder_generator(root))

	@classmethod
	def inorder_generator(cls, tree: Optional[TreeNode]):
		if tree is not None:
			yield from cls.inorder_generator(tree.left)
			yield tree.val
			yield from cls.inorder_generator(tree.right)
